### DISCUSSION TITLE
Build Export Support

### DISCUSSION THREAD

**Discussion by decanTyme** [warengonzaga/css-text-portrait-builder#47] [Brainstorm]
Almost didn't realize I had commented on a maintainers-only issue (#10) so I'm gonna move this here:

> Still thinking of using GulpJS to achieve this. I don't know if Parcel can package the project that can work without a server.

I think we can just export and compress Parcel builds since they do contain the necessary `html` file, then it's up to the user how they will serve/showcase their exported builds. They can just open it manually in a browser if they like, tho when I tried to it didn't work until I manually corrected the `src` paths and removed `type="module"` from the `<script>` tag in the `index.html` because:

![image](https://user-images.githubusercontent.com/74456102/143681817-5b0717be-cbf4-4107-af0b-94b1c85ba65c.png)

But if they serve it with a local server it [should be fine](https://stackoverflow.com/a/10752078/17149207). They can also just directly edit these builds since it should contain all necessary deps, albeit being minified.

It's just a simple manual correction but if you do want this to be as open as possible to the _non-technical_ then it would need have some sort of postbuild script to mitigate the issue above. Or something like [Web Server for Chrome](https://chrome.google.com/webstore/detail/web-server-for-chrome/ofhbbkphhbklhfoeikjpcbhemlocgigb?hl=en). I don't see any other solutions as of yet.

What are your thoughts on this? How would they get their builds?

**Comment by warengonzaga** [comment]
That's the problem with Parcel @decanTyme that's why I'm looking into ways to export builds "as is" without any web server needed. I normally achieve this with GulpJS a simple compressed/minified build and then package it as a zip. I'm looking for the same thing in Parcel well this is my first time to use parcel in my open source project I have normally used GulpJS before.

I'm currently thinking of adding GulpJS to the workflow with Parcel. My mind told me that it might be redundant. I'm really not sure regarding this.

**Reply by decanTyme** [reply]
If it's just a simple correction then npm scripts should be enough for now (e.g. something like `yarn run export`). I'll try my hands on that too, but would it be better if it wasn't minified so they can edit it later?

**Reply by warengonzaga** [reply]
> If it's just a simple correction then npm scripts should be enough for now (e.g. something like yarn run export). I'll try my hands on that too, but would it be better if it wasn't minified so they can edit it later?

I guess, let's add an option for minified and dev versions of export. Let's say export type (minified and dev).

**Reply by decanTyme** [reply]
Have some great news! Finally found a way to build a custom `postbuild` script that can just run in node + yarn. Now we're getting closer to a good export support. All that's left now is to add some scripts. Suggestions as follows: there's `export`, `export:dev` for _non-minified_, and `export:prod`, `export:minified` for _minified_. I'm thinking there should also be something like `export:gzipped` or `export:compressed` since Parcel supports compressors natively.

**Reply by warengonzaga** [reply]
> Have some great news! Finally found a way to build a custom postbuild script that can just run in node + yarn. Now we're getting closer to a good export support. All that's left now is to add some scripts. Suggestions as follows: there's export, export:dev for non-minified, and export:prod, export:minified for minified. I'm thinking there should also be something like export:gzipped or export:compressed since Parcel supports compressors natively.

That's good news! Let's do it! 👌

**Reply by decanTyme** [reply]
I have some (kinda?) bad news. I was so concerned with how to postprocess the html so that it wouldn't need a server to view on export that I forgot that the src js uses import/export, which means Parcel transpiles it to target older js. The transpiled js would then be almost uneditable because of import/export artifacts that most users will not be able to understand (if they wanna edit it after.) I have tried almost everything including having a babel config. I can PR the stuff if you'd like so you can take a look at what I mean. 

Sample compressed export: [export.zip](https://github.com/WarenGonzaga/css-text-portrait-builder/files/7656160/export.zip)

There's also the problem of Yarn not executing pre and post scripts.

**Reply by warengonzaga** [reply]
I don't see any problem with this, since exporting builds means final output. If you want to tweak or edit your build you need to use the builder as a whole.

**Comment by warengonzaga** [comment]
Issue #10 is now open to the public and potential contributors.
